Add simple benchmarks to BloomFilterBuilder with initial performance findings

AlvinKuruvilla opened the pull request:
![insert](https://github.com/user-attachments/assets/89daadf0-d42b-40e3-8e84-51d102209fb3)
There are benchmarks for normal bloom filter operations (building, querying, and insertion). However, this commit's main benchmark of interest is `bloom_filter_insert_benchmark,` which allows us to stress `filter_hash()`. As the flamegraph indicates, we spend roughly 12% of benchmark execution time in `filter_hash()`.

The hope is to reduce this in subsequent PRs/patches.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slatedb/slatedb?pullRequest=250) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slatedb/slatedb?pullRequest=250) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slatedb/slatedb?pullRequest=250) it.</sub>

rodesai commented on Cargo.toml:
```diff
@@ -84,3 +95,6 @@ name = "slatedb"
 
+[[bench]]
+name = "bloom_bench"
+harness = false
```
this is false because we're using criterion?

AlvinKuruvilla replied:
Yep, `harness = false`, disables the default test harness and uses criterion's harness instead

rodesai commented on benches/bloom_bench.rs:
```diff
@@ -0,0 +1,64 @@
+use slatedb::filter::{filter_hash, BloomFilterBuilder};
+fn bloom_filter_insert_benchmark(c: &mut Criterion) {
+    let mut builder = BloomFilterBuilder::new(10); // Set bits_per_key
+    let key_sz = std::mem::size_of::<u32>();
```
we probably want to use a more realistic key size here (4 bytes seems rather small)

AlvinKuruvilla replied:
> we probably want to use a more realistic key size here (4 bytes seems rather small)

Would this be better?
`let key_sz = std::mem::size_of::<u64>();`

rodesai replied:
I think we should make it flexible (e.g. use a `Vec<u8>` to store the key) and default it to something larger. IMO 32 bytes would be reasonable.

criccomini replied:
@AlvinKuruvilla I think this still needs to be addressed. `key_sz` still appears to be 4 bytes.

rodesai commented on benches/bloom_bench.rs:
```diff
@@ -0,0 +1,64 @@
+        })
+    });
+}
+fn bloom_filter_build_benchmark(c: &mut Criterion) {
```
rather than make all the bloom filter types public, could we just add a public filter_bench module to the lib and put these methods in there? They can/should still be called from `bloom_bench.rs`, since it looks like Criterion can't be embedded alongside the source code.

criccomini replied:
@AlvinKuruvilla This still needs to be addressed.

rodesai commented on benches/bloom_bench.rs:
```diff
@@ -0,0 +1,64 @@
+    c.bench_function("bloom_filter insert 10000000 keys", |b| {
+        b.iter(|| {
+            for i in 0..10000000 {
+                let mut bytes = BytesMut::with_capacity(key_sz);
```
I don't think we should be generating the keys inside of the benchmark. We want the benchmark to give us some idea of the throughput of the code under test, which shouldn't include key generation which can be quite expensive. Having a more representative number will help us better understand the magnitude of any improvements we make. I'd suggest pre-generating these outside the bench method.

rodesai commented on benches/bloom_bench.rs:
```diff
@@ -0,0 +1,64 @@
+
+    c.bench_function("bloom_filter insert 10000000 keys", |b| {
+        b.iter(|| {
+            for i in 0..10000000 {
```
For the insert benchmark I don't think we need to insert so many keys. To understand key insert performance a single key would suffice. Criterion will take care of running the benchmarked code over many iterations to get representative data.

rodesai commented on benches/bloom_bench.rs:
```diff
@@ -0,0 +1,64 @@
+            let _filter = black_box(builder.build());
+        })
+    });
+}
```
It would be good to have a benchmark that tests the performance of building the filter for an entire SST - e.g. combining the above two benchmarks into one that both inserts keys and builds the filter. That (alongside querying the filter) is really what we care about optimizing for filters since that's what the calling code does. The specific benchmarks for insert and build are nice if we see that the combined benchmark is spending a lot of time in either place and we want to see if an optimization targeting insert or build worked, but I think what we'd want to measure to produce a reference number (e.g. for regression tests) is building a whole filter.

criccomini replied:
@AlvinKuruvilla Same here.

AlvinKuruvilla commented:
I appreciate the review @rodesai! I'll get to work on these changes and circle back

criccomini commented:
@AlvinKuruvilla Checking in here: Any plans to update this one?

AlvinKuruvilla commented:
Apologies for the delay, @criccomini. I think I'm going be too busy with school for at least the next month. So I most likely won't get a chance to work on this PR. If there is interest from anyone to take it over they definitely can. If not, I can circle back to this issue when I have some more bandwidth

AlvinKuruvilla commented:
Apologies for the delay on this PR, @criccomini; I believe I have addressed most of your comments. 
> rather than make all the bloom filter types public, could we just add a public filter_bench module to the lib and put these methods in there? They can/should still be called from `bloom_bench.rs`, since it looks like Criterion can't be embedded alongside the source code.

Regarding this, I attempted to make a filter_bench module in lib.rs and then adjusted the path in the TOML file to reference the new filter_bench module's mod.rs file. However, I ran into issues with the compiler being unable to find the `filter` module in something like:

`use crate::filter::{filter_hash, BloomFilterBuilder};`

Any other feedback is always appreciated and again sorry for the delay :)

criccomini commented on Cargo.toml:
```diff
@@ -17,6 +17,7 @@ bytemuck = "1.18.0"
 bytes = { version = "1.7.1", features = ["serde"] }
 clap = { version = "4.5", optional = true, features = ["derive"] }
 crc32fast = "1.4.2"
+criterion = { version = "0.5.1", features = ["async_tokio", "html_reports"] }
```
I don't think this is needed. We have it in `dev-dependencies` already.

criccomini reviewed (changes requested):
@AlvinKuruvilla Looks like there are several change requests from Rohan that still need to be addressed. I also left a note on the Cargo.toml file.
